### PR TITLE
Fix crash on RigidBody _direct_state_changed

### LINKED ISSUE #46003: Executing `RigidBody.new()._direct_state_changed(BoxShape.new())` crashes Godot
**Godot version:**
Godot 3.2.4 rc 2

**Issue description:**
Executing
```
RigidBody.new()._direct_state_changed(BoxShape.new())
```

crashes with backtrace
```
scene/3d/physics_body.cpp:453:44: runtime error: member call on null pointer of type 'struct PhysicsDirectBodyState'
scene/3d/physics_body.cpp:453:44: runtime error: member access within null pointer of type 'struct PhysicsDirectBodyState'
handle_crash: Program crashed with signal 11
Dumping the backtrace. Please include this when reporting the bug on https://github.com/godotengine/godot/issues
[1] godots() [0x16f757a] (/mnt/Miecz/godot3.2/platform/x11/crash_handler_x11.cpp:54)
[2] /lib/x86_64-linux-gnu/libc.so.6(+0x46210) [0x7f7e6f8a4210] (??:0)
[3] RigidBody::_direct_state_changed(Object*) (/mnt/Miecz/godot3.2/scene/3d/physics_body.cpp:453)
[4] MethodBind1<Object*>::call(Object*, Variant const**, int, Variant::CallError&) (/mnt/Miecz/godot3.2/./core/method_bind.gen.inc:775 (discriminator 12))
[5] Object::call(StringName const&, Variant const**, int, Variant::CallError&) (/mnt/Miecz/godot3.2/core/object.cpp:919 (discriminator 1))
[6] Variant::call_ptr(StringName const&, Variant const**, int, Variant*, Variant::CallError&) (/mnt/Miecz/godot3.2/core/variant_call.cpp:1129 (discriminator 1))
[7] GDScriptFunction::call(GDScriptInstance*, Variant const**, int, Variant::CallError&, GDScriptFunction::CallState*) (/mnt/Miecz/godot3.2/modules/gdscript/gdscript_function.cpp:1089)
[8] GDScriptInstance::_ml_call_reversed(GDScript*, StringName const&, Variant const**, int) (/mnt/Miecz/godot3.2/modules/gdscript/gdscript.cpp:1269)
[9] GDScriptInstance::call_multilevel_reversed(StringName const&, Variant const**, int) (/mnt/Miecz/godot3.2/modules/gdscript/gdscript.cpp:1278)
[10] Node::_notification(int) (/mnt/Miecz/godot3.2/scene/main/node.cpp:152)
[11] Node::_notificationv(int, bool) (/mnt/Miecz/godot3.2/./scene/main/node.h:46 (discriminator 14))
[12] CanvasItem::_notificationv(int, bool) (/mnt/Miecz/godot3.2/./scene/2d/canvas_item.h:166 (discriminator 3))
[13] Node2D::_notificationv(int, bool) (/mnt/Miecz/godot3.2/./scene/2d/node_2d.h:38 (discriminator 3))
[14] Object::notification(int, bool) (/mnt/Miecz/godot3.2/core/object.cpp:931)
[15] Node::_propagate_ready() (/mnt/Miecz/godot3.2/scene/main/node.cpp:197)
[16] Node::_propagate_ready() (/mnt/Miecz/godot3.2/scene/main/node.cpp:186 (discriminator 2))
[17] Node::_set_tree(SceneTree*) (/mnt/Miecz/godot3.2/scene/main/node.cpp:2560)
[18] SceneTree::init() (/mnt/Miecz/godot3.2/scene/main/scene_tree.cpp:464)
[19] OS_X11::run() (/mnt/Miecz/godot3.2/platform/x11/os_x11.cpp:3628)
[20] godots(main+0x331) [0x16ee467] (/mnt/Miecz/godot3.2/platform/x11/godot_x11.cpp:57)
[21] /lib/x86_64-linux-gnu/libc.so.6(__libc_start_main+0xf3) [0x7f7e6f8850b3] (??:0)
[22] godots(_start+0x2e) [0x16ee07e] (??:?)
```

### PR DESCRIPTION
**Pull request by rafallus**
Fixes: #46003 for master branch

### CODE REVIEW

**Review by madmiraal — CHANGES_REQUESTED** [review]
Three points:
1. As mentioned in https://github.com/godotengine/godot/issues/46003#issuecomment-809984407, unless there is a reason this method is exposed, the better solution would be to unexpose this method.
2. This method also exists in KinematicBody, VehicleBody and the equivalents in 2D; so whatever solution is agreed on, those should be updated too.
3. This will prevent the crash identified in #47485 in debug mode only i.e. when run from the editor.

**Comment by rafallus** [comment]
@madmiraal  
1. I agree that's a decision to be made.
2. Thanks for pointing this out.
3. The comment for the non-debug section says to trust it. So I just leave it as is. I guess for release you are supposed to have already tested your code.

**Comment by pouleyKetchoupp** [comment]
I agree we should just unexpose the method on master. The method binding seems to be used only because of an internal call (see https://github.com/godotengine/godot/issues/46003#issuecomment-810003807).

@rafallus If you're up for refactoring this fix, the proper changes are:
* Modify `body_set_force_integration_callback` in the physics server to use a callable:
https://github.com/godotengine/godot/blob/35066f39ff111ee0e9d4aba534bb08fc2d686156/servers/physics_server_3d.h#L489
```
virtual void body_set_force_integration_callback(RID p_body, const Callable &p_callable, const Variant &p_udata = Variant()) = 0; 
```
* Modify how it's stored in the body inside the physics server:
https://github.com/godotengine/godot/blob/35066f39ff111ee0e9d4aba534bb08fc2d686156/servers/physics_3d/body_3d_sw.h#L129-L133
```
struct ForceIntegrationCallback {
	Callable callable;
	Variant udata;
};
```
* Modify the call to use `Callable::get_object()` and `Callable::call()`:
https://github.com/godotengine/godot/blob/35066f39ff111ee0e9d4aba534bb08fc2d686156/servers/physics_3d/body_3d_sw.cpp#L696-L705
* Make the same changes in Bullet module and Godot Physics 2D
* Use `callable_mp` macro in calls to `body_set_force_integration_callback`
* Remove all the `_direct_state_changed` bindings

**Comment by pouleyKetchoupp** [comment]
@rafallus Thanks for refactoring the PR! I haven't reviewed or tested yet, but for info some checks are failing because of compilation errors in test code:
```
tests/test_physics_2d.cpp: In member function 'RID TestPhysics2DMainLoop::_add_body(PhysicsServer2D::ShapeType, const Transform2D&)':
tests/test_physics_2d.cpp:246:76: error: no matching function for call to 'PhysicsServer2D::body_set_force_integration_callback(RID&, TestPhysics2DMainLoop*, const char [12], RID&)'
  246 |   ps->body_set_force_integration_callback(body, this, "_body_moved", sprite);
      |                                                                            ^
In file included from tests/test_physics_2d.cpp:39:
./servers/physics_server_2d.h:480:15: note: candidate: 'virtual void PhysicsServer2D::body_set_force_integration_callback(RID, const Callable&, const Variant&)'
  480 |  virtual void body_set_force_integration_callback(RID p_body, const Callable &p_callable, const Variant &p_udata = Variant()) = 0;
      |               ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
./servers/physics_server_2d.h:480:15: note:   candidate expects 3 arguments, 4 provided
```
It seems there are some more cases in `test_physics_3d.cpp` as well.

**Comment by rafallus** [comment]
Still there are some tests that haven't passed. I'll check that later

**Comment by rafallus** [comment]
Ok, the PR is ready for review

**Review by pouleyKetchoupp — CHANGES_REQUESTED** [review]
Looks good except for 2 things:

* There are currently errors when starting the editor or game because of `body_set_force_integration_callback` in physics servers has different arguments now. The method binding needs to be updated here:
https://github.com/godotengine/godot/blob/5c8505c25adf9524bd9ea2bc2ad31eb44190ffcc/servers/physics_server_2d.cpp#L650
(and the same in the 3d physics server)

* The documentation for physics servers needs to be updated too:
https://github.com/godotengine/godot/blob/5c8505c25adf9524bd9ea2bc2ad31eb44190ffcc/doc/classes/PhysicsServer2D.xml#L657-L671
(again, for the 3d version as well)

**Review by pouleyKetchoupp — APPROVED** [review]

**Review by pouleyKetchoupp — APPROVED** [review]

File: tests/test_physics_2d.cpp [hunk 1]
```diff
@@ -243,7 +243,7 @@ class TestPhysics2DMainLoop : public MainLoop {
 		Size2 imgsize(5, 5); //vs->texture_get_width(body_shape_data[p_shape].image), vs->texture_get_height(body_shape_data[p_shape].image));
 		vs->canvas_item_add_texture_rect(sprite, Rect2(-imgsize / 2.0, imgsize), body_shape_data[p_shape].image);
 
-		ps->body_set_force_integration_callback(body, this, "_body_moved", sprite);
+		ps->body_set_force_integration_callback(body, callable_mp(this, &TestPhysics2DMainLoop::_body_moved), sprite);
```
Comment (akien-mga) [inline tests/test_physics_2d.cpp:246]: This should let you remove the `ClassDB::bind_method(D_METHOD("_body_moved"), &TestPhysics2DMainLoop::_body_moved);` which was only used for this callback.

(And maybe clear the two commented lines below this which won't be reused as is.)

File: tests/test_physics_3d.cpp [hunk 2]
```diff
@@ -93,7 +93,7 @@ class TestPhysics3DMainLoop : public MainLoop {
 		ps->body_set_param(body, PhysicsServer3D::BODY_PARAM_BOUNCE, 0.0);
 		//todo set space
 		ps->body_add_shape(body, type_shape_map[p_shape]);
-		ps->body_set_force_integration_callback(body, this, "body_changed_transform", mesh_instance);
+		ps->body_set_force_integration_callback(body, callable_mp(this, &TestPhysics3DMainLoop::body_changed_transform), mesh_instance);
```
Comment (akien-mga) [inline tests/test_physics_3d.cpp:96]: You can also remove `ClassDB::bind_method("body_changed_transform", &TestPhysics3DMainLoop::body_changed_transform);` (and the whole `_bind_method` override).

**Review by akien-mga — APPROVED** [review]
Suggested a couple additional cleanups, but looks great overall. Thanks!

**Review by madmiraal — CHANGES_REQUESTED** [review]
It looks good. Just, as suggested below, the original check this PR suggested either needs to be applied consistently or not at all.

**Review by madmiraal — APPROVED** [review]

**Comment by akien-mga** [comment]
Thanks!

**Comment by twocountriestrus** [comment]
Gracias 


Enviado desde Yahoo Mail para iPhone


El viernes, abril 23, 2021, 10:03 a. m., Rémi Verschelde ***@***.***> escribió:




Thanks!

**Comment by pycbouh** [comment]
@twocountriestrus Please don't spam PRs with "Thanks"

**Pull request merged**